cpg-language-go/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/GoEvaluationOrderGraphPass.kt

```diff
@@ -90,13 +88,14 @@ class GoEvaluationOrderGraphPass(ctx: TranslationContext) : EvaluationOrderGraph
         // We need to follow the path from the defer statement to all return statements that are
         // reachable from this point.
         for (defer in defers ?: listOf()) {
-            val paths = defer.followNextEOGEdgesUntilHit { it is ReturnStatement }
+            // TODO(oxisto): This is broken! this returns 8192 paths instead of 4
```

Comment:
All the normal tests seem to pass, but if I try to parse https://github.com/clouditor/clouditor/blob/09e6934d56ed8d1869291efcd4091fe16ddac855/cli/session.go#L155-L166, I get 3145728 items in `paths`, which can be reduced to 2 unique paths, which seem to be correct. But I have no idea why the number of items explodes in this case.

We could also leave it in for now and then revisit the bug in a later PR.

Reply:
I extracted this into a minimum example:

```
package p

type MyStruct struct {
	saved bool
}

func (s *MyStruct) Save() (err error) {
	defer after()

	if err = check(); err != nil {
		return err
	}

	s.saved = true

	return nil
}

func after()       {}
func check() error { return nil }

func something() {}
```

This simple example seems to confuse followNextEOGEdgesUntilHit, producing 34816 paths instead of 2. Interesting enough, the `ControlFlowSensitiveDFGPass` which gets executed afterwards also seems to have problems, in total the EOG pass takes 9655 ms and the ControlFlowSensitiveDFGPass 45422 ms.

If I remove the line `s.saved = true`, it gets reduced to 4096 paths (there are still only 2 valid ones), but at least it gets executed in less than a second.

Reply:
Aaaah. This was a rather stupid bug; the logic for the defer handling was called twice because method declarations in Go are iterated twice in the EOG handler. This is due to a workaround where we add the methods of a struct in a file to both the current namespace as well as their record.

Reply:
Well, if the EOG is a bit weird, the `ControlFlowSensitiveDFGPass` is likely to run into trouble. This also happens frequently when generating tests with the Fluent DSL. Is this issue resolved now?

Reply:
Seems so. Yes